### ISSUE TITLE
Can't use CUDA and torch for variational quantum circuits

### ISSUE THREAD

**Issue by dnnagy** [PennyLaneAI/pennylane#787]
I just started to work with PennyLane, and I'm trying to create a variational quantum classifier, which uses pytorch for learning the parameters of the circuit. On CPU it works well, but when I try to use CUDA for parameter learning, it fails.

* *System information:*
torch version: 1.6.0
CUDA version: 9.1.85
Name: PennyLane
Version: 0.11.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/dnagy/anaconda3/envs/pt11/lib/python3.6/site-packages
Requires: scipy, autograd, networkx, toml, appdirs, numpy, semantic-version
Required-by: PennyLane-SF, PennyLane-Forest
Platform info:           Linux-4.15.0-88-generic-x86_64-with-debian-buster-sid
Python version:          3.6.10
Numpy version:           1.18.5
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.11.0)
- default.qubit (PennyLane-0.11.0)
- default.qubit.autograd (PennyLane-0.11.0)
- default.qubit.tf (PennyLane-0.11.0)
- default.tensor (PennyLane-0.11.0)
- default.tensor.tf (PennyLane-0.11.0)
- strawberryfields.fock (PennyLane-SF-0.11.0)
- strawberryfields.gaussian (PennyLane-SF-0.11.0)
- strawberryfields.remote (PennyLane-SF-0.11.0)
- forest.numpy_wavefunction (PennyLane-Forest-0.9.0)
- forest.qpu (PennyLane-Forest-0.9.0)
- forest.qvm (PennyLane-Forest-0.9.0)
- forest.wavefunction (PennyLane-Forest-0.9.0)

#### Source code and tracebacks

```python
import strawberryfields as sf
import torch
import torch.nn as nn
from torch.nn import functional as F

import pennylane as qml
import pennylane_sf

device = qml.device('strawberryfields.fock', cutoff_dim=10, wires=2)

alpha = torch.tensor(0.5, requires_grad=True).to('cuda:0')
beta = torch.tensor(0.65, requires_grad=True).to('cuda:0')
theta = torch.tensor(0.45, requires_grad=True).to('cuda:0')
phi = torch.tensor(0.25, requires_grad=True).to('cuda:0')

@qml.qnode(device, interface='torch')
def cv_test(alpha, beta, theta, phi):
    qml.Displacement(alpha, 0.0, wires=[0])
    qml.Displacement(beta, 0.0, wires=[1])
    qml.Beamsplitter(theta, phi, wires=[w for w in [0,1]])
    return [qml.expval(qml.NumberOperator(w)) for w in [0,1]]

result = cv_test(alpha, beta, theta, phi)

loss = F.cross_entropy(result.unsqueeze(0), torch.tensor([1], dtype=torch.int64).to('cuda:0'))

loss.backward()
````
Traceback:
```bash
/home/dnagy/anaconda3/envs/pt11/lib/python3.6/site-packages/strawberryfields/program.py:550: UserWarning: The circuit consists of 2 disconnected components.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dnagy/anaconda3/envs/pt11/lib/python3.6/site-packages/torch/tensor.py", line 185, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/dnagy/anaconda3/envs/pt11/lib/python3.6/site-packages/torch/autograd/__init__.py", line 127, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: CUDA error: an illegal memory access was encountered
```

**Comment by nquesada** [comment]
Hi @dnnagy . Thanks for the bug report. It is not very obvious what is going on, but it might be a bug in torch, especially since the code works fine when run using CPUs. Here https://discuss.pytorch.org/t/runtimeerror-cuda-error-an-illegal-memory-access-was-encountered/47999/5 they report similar issues.

**Comment by josh146** [comment]
Since the cause of this bug was not determined, I will close this issue for now @dnnagy. Feel free to re-open, or create a new issue, if you have more information/details!
